In [1]:
import pandas as pd

path = "data/ai_agent_telemetry_2026.csv"

df = pd.read_csv(path)


In [2]:
df.columns

Index(['session_id', 'timestamp', 'company_sector', 'company_size',
       'cloud_region', 'agent_framework', 'agent_architecture',
       'foundation_model', 'task_domain', 'task_complexity',
       'developer_experience', 'prompt_summary', 'input_tokens',
       'output_tokens', 'reasoning_tokens', 'context_window_kb',
       'iteration_steps', 'tools_invoked_count', 'tool_error_rate',
       'latency_seconds', 'total_cost_usd', 'agent_confidence_score',
       'network_egress_kb', 'human_intervention_required',
       'intervention_reason', 'task_status', 'human_satisfaction_score',
       'post_task_bug_detected', 'error_log_snippet'],
      dtype='str')

In [3]:
df.head()

,session_id,timestamp,company_sector,company_size,cloud_region,agent_framework,agent_architecture,foundation_model,task_domain,task_complexity,...,latency_seconds,total_cost_usd,agent_confidence_score,network_egress_kb,human_intervention_required,intervention_reason,task_status,human_satisfaction_score,post_task_bug_detected,error_log_snippet
0,agt_1fe1dc66000000,2025-10-22 18:35:57,Legal & Compliance,Startup (<50),US-West,CrewAI,Single-Agent ReAct,DeepSeek-V3,Financial Data Extraction & Reconciliation,Medium,...,92.97,0.0097,0.653,222.5,0,NaN,Completed_Autonomous_Success,4.0,0,NaN
1,agt_0bea3db3000001,2026-06-27 22:52:56,Industrial Automation,Startup (<50),US-East,AutoGen v0.4,Multi-Agent Swarm (Consensus),DeepSeek-V3,Customer Dispute Autopilot,Medium,...,82.84,0.0071,0.788,248.4,1,Security Guardrail Violation,Completed_With_Human_Correction,4.0,0,AssertionError: Generated unit tests failed (3...
2,agt_3362035c000002,2025-10-13 13:00:25,E-Commerce & Retail,Mid-Market (50-500),EU-West (Ireland),Semantic Kernel,Single-Agent ReAct,GPT-4.5-Preview,Full-Stack Feature Engineering,High,...,212.03,0.9204,0.667,407.4,0,NaN,Completed_Autonomous_Success,NaN,0,NaN
3,agt_2ef5950e000003,2025-11-13 20:48:28,FinTech,Mid-Market (50-500),AP-East (Tokyo),Custom In-House SDK,Plan-and-Solve (Hierarchical),Claude-3.7-Sonnet,Full-Stack Feature Engineering,Low,...,105.84,0.2196,0.441,365.1,0,NaN,Completed_Autonomous_Success,5.0,0,NaN
4,agt_3b63f46a000004,2026-06-15 11:44:04,E-Commerce & Retail,Mega-Corp (5000+),US-East,LlamaIndex Workflows,Multi-Agent Swarm (Consensus),Claude-3.7-Sonnet,Vulnerability Patching & Security Audit,High,...,151.75,0.2725,0.794,413.3,0,NaN,Completed_Autonomous_Success,4.0,0,NaN


In [4]:
df[df["human_intervention_required"] == 1].isna().sum()

session_id                         0
timestamp                          0
company_sector                     0
company_size                       0
cloud_region                       0
agent_framework                    0
agent_architecture                 0
foundation_model                   0
task_domain                        0
task_complexity                    0
developer_experience               0
prompt_summary                     0
input_tokens                       0
output_tokens                      0
reasoning_tokens               13751
context_window_kb                  0
iteration_steps                    0
tools_invoked_count                0
tool_error_rate                    0
latency_seconds                    0
total_cost_usd                     0
agent_confidence_score          1407
network_egress_kb                714
human_intervention_required        0
intervention_reason             1119
task_status                        0
human_satisfaction_score        7909
p

In [5]:
df[df["human_intervention_required"] == 1]["intervention_reason"].value_counts()

intervention_reason
Tool Execution Permission Denied                   5020
Hallucination / Semantic Inconsistency             3983
Infinite Loop / Circular Reasoning                 3053
Context Overflow / Degraded Memory                 2199
Security Guardrail Violation                       1613
Token Budget Exceeded                              1266
Ambiguous Specification / Clarification Request     908
Name: count, dtype: int64

In [6]:
pd.crosstab(
    df["human_intervention_required"],
    df["intervention_reason"].isna()
)

intervention_reason,False,True
human_intervention_required,,
0,0,160839
1,18042,1119


In [8]:
df["task_status"].value_counts()

task_status
Completed_Autonomous_Success       149995
Silent_Failure_Uncaught_Bug          9566
Failed_User_Aborted                  6707
Completed_With_Human_Correction      5335
Terminated_Hallucination             3949
Terminated_Looping                   3030
Timeout_Aborted                      1418
Name: count, dtype: int64

In [9]:
for col in df.columns:
    print(col)
    print(df[col].value_counts())
    print()

session_id
session_id
agt_1fe1dc66000000    1
agt_0bea3db3000001    1
agt_3362035c000002    1
agt_2ef5950e000003    1
agt_3b63f46a000004    1
                     ..
agt_1195689602bf1b    1
agt_000a0ccb02bf1c    1
agt_30524a4102bf1d    1
agt_2a4a7f1c02bf1e    1
agt_2eeedadc02bf1f    1
Name: count, Length: 180000, dtype: int64

timestamp
timestamp
2026-01-26 11:38:25    2
2026-04-26 12:58:07    2
2025-12-27 18:51:36    2
2025-11-20 02:28:22    2
2025-11-02 14:24:32    2
                      ..
2026-06-09 12:06:59    1
2025-11-01 22:46:26    1
2025-10-28 10:51:59    1
2025-12-21 12:48:38    1
2025-11-11 05:08:40    1
Name: count, Length: 179405, dtype: int64

company_sector
company_sector
FinTech                    38841
Cloud SaaS                 31813
Healthcare & Biotech       28294
CyberSecurity              26784
E-Commerce & Retail        21339
EdTech                     12464
Legal & Compliance         10755
Industrial Automation       7010
fintech                      303
 FinTe

In [10]:
df["timestamp"].str[:4].value_counts()

timestamp
2026    116959
2025     59477
01/2       151
02/1       144
04/2       141
10/2       140
10/1       136
12/0       135
03/2       134
02/2       134
11/2       129
01/1       129
05/1       126
11/1       126
06/1       125
03/1       124
12/1       124
11/0       123
06/2       123
06/0       122
05/0       122
02/0       121
10/0       120
01/0       119
04/1       113
05/2       113
12/2       111
03/0       108
04/0       102
05/3        30
03/3        28
12/3        27
01/3        25
10/3        24
11/3        15
04/3        11
06/3         9
Name: count, dtype: int64

In [11]:
df[~df["timestamp"].str.match(r"^\d{4}-\d{2}-\d{2}")]["timestamp"].head(20)

154     02/03/2026 08:04
185     01/07/2026 15:26
209     11/27/2025 07:27
261     10/18/2025 18:03
276     06/05/2026 16:49
277     04/18/2026 14:56
298     12/05/2025 01:04
327     12/08/2025 21:11
380     03/28/2026 03:12
582     10/09/2025 06:38
591     05/16/2026 05:10
633     02/18/2026 18:45
639     11/20/2025 18:07
690     11/22/2025 19:05
718     01/23/2026 14:48
742     01/17/2026 09:36
768     10/12/2025 20:08
778     06/03/2026 00:29
930     11/07/2025 18:07
1248    02/07/2026 20:34
Name: timestamp, dtype: str

In [12]:
import sys
print(sys.executable)

e:\profile\learning\projects\ml-proj1\.venv\Scripts\python.exe


In [13]:
df[df["timestamp"].str.contains(r"[+-]\d{2}:?\d{2}|Z$", regex=True, na=False)]["timestamp"].head(20)

6      2025-10-12T05:25:14Z
22     2026-02-17T13:11:25Z
26     2026-02-21T06:01:12Z
48     2025-11-28T11:25:57Z
76     2025-11-18T23:38:21Z
86     2026-05-05T16:33:23Z
91     2025-10-11T15:02:20Z
98     2025-12-10T10:17:16Z
100    2026-05-30T08:04:06Z
103    2025-12-16T15:10:06Z
104    2026-05-11T21:10:45Z
162    2026-01-23T02:16:53Z
170    2026-06-05T02:20:40Z
243    2026-01-28T07:23:26Z
245    2026-01-22T15:52:12Z
278    2026-06-09T21:27:12Z
303    2025-11-06T13:30:42Z
318    2025-11-11T09:34:18Z
324    2026-01-20T06:09:33Z
328    2026-03-04T16:15:18Z
Name: timestamp, dtype: str

In [14]:
df["timestamp"].head(20).tolist()

['2025-10-22 18:35:57',
 '2026-06-27 22:52:56',
 '2025-10-13 13:00:25',
 '2025-11-13 20:48:28',
 '2026-06-15 11:44:04',
 '2026-06-16 05:06:29',
 '2025-10-12T05:25:14Z',
 '2025-10-30 21:02:20',
 '2026-04-10 20:53:32',
 '2025-11-10 21:05:21',
 '2025-10-19 16:52:22',
 '2026-03-08 03:28:12',
 '2026-04-24 12:56:12',
 '2026-01-08 23:06:23',
 '2026-03-26 20:59:27',
 '2026-04-01 05:36:35',
 '2026-05-19 19:19:55',
 '2026-05-25 09:38:45',
 '2025-12-18 18:03:32',
 '2025-12-17 04:28:50']

In [15]:
normal_mask = df["timestamp"].str.match(
    r"^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$",
    na=False
)
normal_timstamps = df.loc[normal_mask, "timestamp"]



In [16]:
print(normal_mask.sum())

167464


In [17]:
normal2_mask = df["timestamp"].str.match(
    r"^\d{2}/\d{2}/\d{4} \d{2}:\d{2}$",
    na=False
)

normal2_timestamps = df.loc[normal2_mask, "timestamp"]

print(normal2_mask.sum())

3564


In [18]:
other_timestamp = df["timestamp"].str.match(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$", na=False)
print(other_timestamp.sum())
###2025-10-12T05:25:14Z

8972


In [19]:
other_timestamps = df.loc[~(normal_mask | normal2_mask), "timestamp"]
print(other_timestamps.count())

8972


In [20]:
normal_timstamps = pd.to_datetime(normal_timstamps, format="%Y-%m-%d %H:%M:%S", errors="coerce")
normal2_timstamps = pd.to_datetime(normal2_timestamps, format="%m/%d/%Y %H:%M", errors="coerce")
other_timestamps = pd.to_datetime(
    df.loc[other_timestamp, "timestamp"],
    format="%Y-%m-%dT%H:%M:%SZ",
    utc=True
)


In [21]:
print(normal_timstamps.count() + normal2_timstamps.count() + other_timestamps.count())

180000


In [22]:
print(normal_timstamps.head())
print(normal2_timstamps.head())
print(other_timestamps.head())

0   2025-10-22 18:35:57
1   2026-06-27 22:52:56
2   2025-10-13 13:00:25
3   2025-11-13 20:48:28
4   2026-06-15 11:44:04
Name: timestamp, dtype: datetime64[us]
154   2026-02-03 08:04:00
185   2026-01-07 15:26:00
209   2025-11-27 07:27:00
261   2025-10-18 18:03:00
276   2026-06-05 16:49:00
Name: timestamp, dtype: datetime64[us]
6    2025-10-12 05:25:14+00:00
22   2026-02-17 13:11:25+00:00
26   2026-02-21 06:01:12+00:00
48   2025-11-28 11:25:57+00:00
76   2025-11-18 23:38:21+00:00
Name: timestamp, dtype: datetime64[us, UTC]


In [23]:
other_timestamps.dt.strftime("%z").value_counts()

timestamp
+0000    8972
Name: count, dtype: int64

Timestamp cleaning: The dataset contains 8,972 timestamps with explicit UTC timezone information, while the remaining 171,028 timestamps are timezone-naive. Since the dataset documentation does not specify a timezone convention for the naive timestamps and our analysis focuses on date/time patterns rather than absolute time across timezones, we standardize the timestamps to a consistent timezone-naive datetime format.

In [24]:
other_timestamps = other_timestamps.dt.tz_convert(None)
other_timestamps.head()

6    2025-10-12 05:25:14
22   2026-02-17 13:11:25
26   2026-02-21 06:01:12
48   2025-11-28 11:25:57
76   2025-11-18 23:38:21
Name: timestamp, dtype: datetime64[us]

In [25]:
df["timestamp"] = pd.concat([normal_timstamps, normal2_timstamps, other_timestamps])

In [26]:
df.columns

Index(['session_id', 'timestamp', 'company_sector', 'company_size',
       'cloud_region', 'agent_framework', 'agent_architecture',
       'foundation_model', 'task_domain', 'task_complexity',
       'developer_experience', 'prompt_summary', 'input_tokens',
       'output_tokens', 'reasoning_tokens', 'context_window_kb',
       'iteration_steps', 'tools_invoked_count', 'tool_error_rate',
       'latency_seconds', 'total_cost_usd', 'agent_confidence_score',
       'network_egress_kb', 'human_intervention_required',
       'intervention_reason', 'task_status', 'human_satisfaction_score',
       'post_task_bug_detected', 'error_log_snippet'],
      dtype='str')

In [27]:
df["timestamp"].dtype

dtype('<M8[us]')

In [28]:
df["timestamp"].isna().sum()

np.int64(0)

In [29]:
df["timestamp"].min(), df["timestamp"].max()

(Timestamp('2025-10-01 00:00:46'), Timestamp('2026-06-30 23:58:08'))

In [30]:
df.drop(columns="session_id", inplace=True)
df.columns

Index(['timestamp', 'company_sector', 'company_size', 'cloud_region',
       'agent_framework', 'agent_architecture', 'foundation_model',
       'task_domain', 'task_complexity', 'developer_experience',
       'prompt_summary', 'input_tokens', 'output_tokens', 'reasoning_tokens',
       'context_window_kb', 'iteration_steps', 'tools_invoked_count',
       'tool_error_rate', 'latency_seconds', 'total_cost_usd',
       'agent_confidence_score', 'network_egress_kb',
       'human_intervention_required', 'intervention_reason', 'task_status',
       'human_satisfaction_score', 'post_task_bug_detected',
       'error_log_snippet'],
      dtype='str')

In [31]:
df["company_sector"].value_counts()

company_sector
FinTech                    38841
Cloud SaaS                 31813
Healthcare & Biotech       28294
CyberSecurity              26784
E-Commerce & Retail        21339
EdTech                     12464
Legal & Compliance         10755
Industrial Automation       7010
fintech                      303
 FinTech                     265
 Cloud SaaS                  238
healthcare & biotech         230
cloud saas                   226
 CyberSecurity               223
cybersecurity                219
 Healthcare & Biotech        216
e-commerce & retail          172
 E-Commerce & Retail         138
 EdTech                       96
edtech                        92
 Legal & Compliance           89
legal & compliance            88
 Industrial Automation        63
industrial automation         42
Name: count, dtype: int64

In [32]:
df["company_sector"] = (df["company_sector"].str.strip().str.lower())

In [33]:
df["company_sector"].value_counts()

company_sector
fintech                  39409
cloud saas               32277
healthcare & biotech     28740
cybersecurity            27226
e-commerce & retail      21649
edtech                   12652
legal & compliance       10932
industrial automation     7115
Name: count, dtype: int64

In [34]:
df.columns

Index(['timestamp', 'company_sector', 'company_size', 'cloud_region',
       'agent_framework', 'agent_architecture', 'foundation_model',
       'task_domain', 'task_complexity', 'developer_experience',
       'prompt_summary', 'input_tokens', 'output_tokens', 'reasoning_tokens',
       'context_window_kb', 'iteration_steps', 'tools_invoked_count',
       'tool_error_rate', 'latency_seconds', 'total_cost_usd',
       'agent_confidence_score', 'network_egress_kb',
       'human_intervention_required', 'intervention_reason', 'task_status',
       'human_satisfaction_score', 'post_task_bug_detected',
       'error_log_snippet'],
      dtype='str')

In [ ]:
df["company_size"].value_counts()
#clean already

company_size
Mid-Market (50-500)      63201
Enterprise (500-5000)    53932
Startup (<50)            35919
Mega-Corp (5000+)        26948
Name: count, dtype: int64

In [ ]:
df["cloud_region"].value_counts()
#clean already

cloud_region
US-East                54058
US-West                45355
EU-West (Frankfurt)    26909
EU-West (Ireland)      21421
AP-South (Mumbai)      14288
AP-East (Tokyo)        12554
SA-East (Sao Paulo)     5415
Name: count, dtype: int64

In [ ]:
df["agent_framework"].value_counts()
#clean already

agent_framework
LangGraph               50500
AutoGen v0.4            39465
CrewAI                  32387
LlamaIndex Workflows    25027
Custom In-House SDK     21584
Semantic Kernel         11037
Name: count, dtype: int64

In [41]:
df["agent_architecture"].value_counts()
#clean already

agent_architecture
Single-Agent ReAct                           57509
Plan-and-Solve (Hierarchical)                43041
Multi-Agent Swarm (Consensus)                32345
Reflection & Self-Correction                 25389
Tool-Calling Directed Acyclic Graph (DAG)    21716
Name: count, dtype: int64

In [43]:
df["foundation_model"].value_counts()
#clean already

foundation_model
Claude-3.7-Sonnet         46944
GPT-4.5-Preview           35899
Gemini-1.5-Pro            27091
Claude-3.5-Haiku          21729
DeepSeek-V3               19725
Llama-3.3-70B-Instruct    14358
Qwen-2.5-Coder-32B         8873
OpenAI-o3-mini             5381
Name: count, dtype: int64

In [46]:
df["task_domain"].value_counts()

task_domain
Full-Stack Feature Engineering                43649
Automated Code Refactoring                    28540
Vulnerability Patching & Security Audit       25260
Financial Data Extraction & Reconciliation    23609
Customer Dispute Autopilot                    21461
Clinical Trial Protocol Synthesis             14268
Contract Clause Compliance Check              12422
DevOps Infrastructure-as-Code Deploy          10791
Name: count, dtype: int64

In [50]:
for col in df.columns:
    print(df[col].value_counts())
    print(df[col].value_counts().sum())
    print()

timestamp
2026-01-26 11:38:25    2
2026-04-26 12:58:07    2
2025-12-27 18:51:36    2
2025-11-20 02:28:22    2
2025-11-02 14:24:32    2
                      ..
2026-06-09 12:06:59    1
2025-11-01 22:46:26    1
2025-10-28 10:51:59    1
2025-12-21 12:48:38    1
2025-11-11 05:08:40    1
Name: count, Length: 179334, dtype: int64
180000

company_sector
fintech                  39409
cloud saas               32277
healthcare & biotech     28740
cybersecurity            27226
e-commerce & retail      21649
edtech                   12652
legal & compliance       10932
industrial automation     7115
Name: count, dtype: int64
180000

company_size
Mid-Market (50-500)      63201
Enterprise (500-5000)    53932
Startup (<50)            35919
Mega-Corp (5000+)        26948
Name: count, dtype: int64
180000

cloud_region
US-East                54058
US-West                45355
EU-West (Frankfurt)    26909
EU-West (Ireland)      21421
AP-South (Mumbai)      14288
AP-East (Tokyo)        12554
SA-East (S

most of the columns are clean as of now with basic eda check. 
these columns need to be investigated further for cleaning and check for potential leakage.
1. prompt_summary
2. error_log_snippet


3. intervention_reason
4. task_status
5. human_satisfaction_score
6. post_task_bug_detection


In [57]:
df["prompt_summary"].nunique()

160

In [58]:
df["prompt_summary"].describe()

count                                                180000
unique                                                  160
top       Implement real-time streaming WebSocket endpoi...
freq                                                   3907
Name: prompt_summary, dtype: object

In [59]:
df["prompt_summary"].unique()

<StringArray>
[         'Extract structured JSON line items from unformatted commercial invoice PDFs [Routine Scheduled Run]',
              'Triage chargeback dispute and aggregate shipping verification telemetry [Routine Scheduled Run]',
                      'Implement real-time streaming WebSocket endpoint with Redis pub/sub [Production Hotfix]',
              'Build OAuth2 multi-tenant authentication handler and session revoke API [Routine Scheduled Run]',
 'Inspect container manifest for privilege escalation vectors and unpinned base images [Routine Scheduled Run]',
                             'Parse SEC 10-K quarterly filings to extract EBITDA, capex, and segment footnotes',
                 'Implement idempotent Stripe webhook handler with exponential retry queue [Production Hotfix]',
                 'Synthesize clinical trial inclusion criteria for diabetic ketoacidosis cohort [High Urgency]',
      'Standardize observational patient cohort demographics for CDISC SDTM submis

In [60]:
pd.crosstab(df["prompt_summary"].str.extract(r"\[(.*?)\]")[0],
            df["human_intervention_required"],
            normalize="index",)

human_intervention_required,0,1
0,,
Experimental Sandbox,0.894444,0.105556
High Urgency,0.890322,0.109678
Production Hotfix,0.892174,0.107826
Routine Scheduled Run,0.894833,0.105167


In [61]:
pd.crosstab(df['prompt_summary'].str.extract(r"(.*?)(?:\[(.*?)\])")[0],
            df["human_intervention_required"],
            normalize="index",)

human_intervention_required,0,1
0,,
Audit smart contract for reentrancy and integer overflow vulnerabilities,0.894785,0.105215
Authenticate tier-2 support ticket for account credential takeover attempt,0.885936,0.114064
Build OAuth2 multi-tenant authentication handler and session revoke API,0.900708,0.099292
Check vendor SLA terms for penalties on 99.9% uptime violation,0.893683,0.106317
Configure Prometheus alerts and OpenTelemetry trace exporter for microservices,0.895780,0.104220
Create responsive analytics dashboard view with GraphQL subscriptions,0.891300,0.108700
Cross-reference adverse drug event reports against FDA MedWatch taxonomy,0.885733,0.114267
Decompose bloated React class components into custom hooks and memoized views,0.896242,0.103758
Draft tailored resolution email for enterprise service outage compensation request,0.898164,0.101836


couldnt decide on whether there is any importance between the task and therefore not able to convert the text into some numerical vaule. decision on this column will be made after seeing the correlation between columns and especially with the human intervention.

In [64]:
df["error_log_snippet"].nunique()

9

In [65]:
df["error_log_snippet"].unique()

<StringArray>
[                                                                                                                    nan,
                                        'AssertionError: Generated unit tests failed (3/14 tests errored on edge cases)',
                           'PermissionError: Action 'delete_production_s3_bucket' blocked by IAM policy governance rule',
                           'TimeoutError: Tool execution exceeded 120.0s threshold while running docker container build',
 'RuntimeWarning: Agent entered cyclical loop across nodes [plan -> generate -> test -> plan]. Recursion limit reached.',
                                   'ConnectionResetError: Remote sandbox terminated abruptly (Exit code 137: OOMKilled)',
                                      'ERROR 429: RateLimitError - tokens per minute quota exceeded on provider gateway',
                              'SecurityException: Prompt injection heuristic triggered in retrieved document chunk #418',
          

In [68]:
pd.crosstab(df["error_log_snippet"],
            df["human_intervention_required"] == 1,
            normalize="index",)

human_intervention_required,False,True
error_log_snippet,,
AssertionError: Generated unit tests failed (3/14 tests errored on edge cases),0.450000,0.550000
ConnectionResetError: Remote sandbox terminated abruptly (Exit code 137: OOMKilled),0.437546,0.562454
ERROR 429: RateLimitError - tokens per minute quota exceeded on provider gateway,0.443235,0.556765
JSONDecodeError: Expecting value: line 1 column 1 (char 0) from LLM unstructured response,0.455991,0.544009
PermissionError: Action 'delete_production_s3_bucket' blocked by IAM policy governance rule,0.439957,0.560043
RuntimeWarning: Agent entered cyclical loop across nodes [plan -> generate -> test -> plan]. Recursion limit reached.,0.442155,0.557845
SecurityException: Prompt injection heuristic triggered in retrieved document chunk #418,0.441937,0.558063
TimeoutError: Tool execution exceeded 120.0s threshold while running docker container build,0.454115,0.545885
ValueError: Schema validation failed on tool output 'git_checkout_branch': unexpected key 'commit_sha',0.440364,0.559636


will decide later on this feature after seeing the correlation

In [71]:
df["intervention_reason"].unique()

<StringArray>
[                                              nan,
                    'Security Guardrail Violation',
                'Tool Execution Permission Denied',
                           'Token Budget Exceeded',
          'Hallucination / Semantic Inconsistency',
              'Infinite Loop / Circular Reasoning',
 'Ambiguous Specification / Clarification Request',
              'Context Overflow / Degraded Memory']
Length: 8, dtype: str

In [72]:
columns_to_check = ["prompt_summary","error_log_snippet","intervention_reason","task_status","human_satisfaction_score","post_task_bug_detected"] 
# 1. prompt_summary
# 2. error_log_snippet


# 3. intervention_reason
# 4. task_status
# 5. human_satisfaction_score
# 6. post_task_bug_detected

In [73]:
for col in columns_to_check:
    print(f"Column: {col}")
    print(df[col].value_counts())
    print(df[col].value_counts().sum())
    print(pd.crosstab(df[col], df["human_intervention_required"], normalize="index"))
    print()

Column: prompt_summary
prompt_summary
Implement real-time streaming WebSocket endpoint with Redis pub/sub [Routine Scheduled Run]           3907
Build OAuth2 multi-tenant authentication handler and session revoke API [Routine Scheduled Run]       3881
Create responsive analytics dashboard view with GraphQL subscriptions [Routine Scheduled Run]         3876
Implement idempotent Stripe webhook handler with exponential retry queue [Routine Scheduled Run]      3849
Migrate deprecated ORM queries to SQLAlchemy 2.0 connection pool syntax [Routine Scheduled Run]       2553
                                                                                                      ... 
Configure Prometheus alerts and OpenTelemetry trace exporter for microservices [High Urgency]          405
Generate Kubernetes Helm chart and GitHub Actions CI/CD deployment pipeline                            399
Generate Kubernetes Helm chart and GitHub Actions CI/CD deployment pipeline [High Urgency]             399

The latter four columns are potential leakage features and will be excluded from the initial feature set, subject to confirmation of their availability relative to the prediction point.

In [77]:
columns_to_check 

['prompt_summary',
 'error_log_snippet',
 'intervention_reason',
 'task_status',
 'human_satisfaction_score',
 'post_task_bug_detected']

In [75]:
for col in df.columns:
    print(f"Column: {col}")
    print(df[col].isna().sum())

Column: timestamp
0
Column: company_sector
0
Column: company_size
0
Column: cloud_region
0
Column: agent_framework
0
Column: agent_architecture
0
Column: foundation_model
0
Column: task_domain
0
Column: task_complexity
0
Column: developer_experience
0
Column: prompt_summary
0
Column: input_tokens
0
Column: output_tokens
0
Column: reasoning_tokens
120713
Column: context_window_kb
0
Column: iteration_steps
0
Column: tools_invoked_count
0
Column: tool_error_rate
0
Column: latency_seconds
0
Column: total_cost_usd
0
Column: agent_confidence_score
13096
Column: network_egress_kb
6643
Column: human_intervention_required
0
Column: intervention_reason
161958
Column: task_status
0
Column: human_satisfaction_score
75614
Column: post_task_bug_detected
0
Column: error_log_snippet
155255


In [78]:
pd.crosstab(
    df["error_log_snippet"].isna(),
    df["human_intervention_required"],
    normalize="index"
)

human_intervention_required,0,1
error_log_snippet,,
False,0.445019,0.554981
True,0.965038,0.034962


In [80]:
pd.crosstab(
    
    df["human_intervention_required"],
    df["error_log_snippet"].isna(),
    normalize="index"
)

error_log_snippet,False,True
human_intervention_required,,
0,0.068466,0.931534
1,0.716716,0.283284


In [79]:
df["error_log_snippet"].unique()

<StringArray>
[                                                                                                                    nan,
                                        'AssertionError: Generated unit tests failed (3/14 tests errored on edge cases)',
                           'PermissionError: Action 'delete_production_s3_bucket' blocked by IAM policy governance rule',
                           'TimeoutError: Tool execution exceeded 120.0s threshold while running docker container build',
 'RuntimeWarning: Agent entered cyclical loop across nodes [plan -> generate -> test -> plan]. Recursion limit reached.',
                                   'ConnectionResetError: Remote sandbox terminated abruptly (Exit code 137: OOMKilled)',
                                      'ERROR 429: RateLimitError - tokens per minute quota exceeded on provider gateway',
                              'SecurityException: Prompt injection heuristic triggered in retrieved document chunk #418',
          

error_log_snippet has a very strong relationship with the target, but its temporal availability must be established before deciding whether it is a valid feature or leakage.